src/Features/Core/Portable/MoveToNamespace/AbstractMoveToNamespaceService.cs

```diff
@@ -0,0 +1,224 @@
+                return new MoveToNamespaceAnalysisResult(document, node, @namespace, namespaces.ToImmutableArray(), MoveToNamespaceAnalysisResult.ContainerType.NamedType);
+            }
+
+            return MoveToNamespaceAnalysisResult.Invalid;
+        }
+
+        private bool ContainsNamespaceDeclaration(SyntaxNode node)
+            => node.DescendantNodes().OfType<TNamespaceDeclarationSyntax>().Any();
+
+        private static bool ContainsMultipleNamespaceInSpine(SyntaxNode node)
+            => node.AncestorsAndSelf().OfType<TNamespaceDeclarationSyntax>().Count() > 1;
+
+        public Task<MoveToNamespaceResult> MoveToNamespaceAsync(
+            MoveToNamespaceAnalysisResult analysisResult,
+            string targetNamespace,
+            CancellationToken cancellationToken)
+        {
+            if (!analysisResult.CanPerform)
+            {
+                return Task.FromResult(MoveToNamespaceResult.Failed);
+            }
+
+            switch (analysisResult.Container)
+            {
+                case MoveToNamespaceAnalysisResult.ContainerType.Namespace:
+                    return MoveItemsInNamespaceAsync(analysisResult.Document, analysisResult.SyntaxNode, targetNamespace, cancellationToken);
+                case MoveToNamespaceAnalysisResult.ContainerType.NamedType:
+                    return MoveTypeToNamespace(analysisResult.Document, analysisResult.SyntaxNode, targetNamespace, cancellationToken);
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private static async Task<MoveToNamespaceResult> MoveItemsInNamespaceAsync(
+            Document document,
+            SyntaxNode container,
+            string targetNamespace,
+            CancellationToken cancellationToken)
+        {
+            var changeNamespaceService = document.GetLanguageService<IChangeNamespaceService>();
+            if (changeNamespaceService == null)
+            {
+                return MoveToNamespaceResult.Failed;
+            }
+
+            var changedSolution = await changeNamespaceService.ChangeNamespaceAsync(
+                document,
+                container,
+                targetNamespace,
+                cancellationToken).ConfigureAwait(false);
+
+            return new MoveToNamespaceResult(changedSolution, document.Id);
+        }
+
+        private static async Task<MoveToNamespaceResult> MoveTypeToNamespace(
+            Document document,
+            SyntaxNode container,
+            string targetNamespace,
+            CancellationToken cancellationToken)
+        {
+            var moveTypeService = document.GetLanguageService<IMoveTypeService>();
+            if (moveTypeService == null)
+            {
+                return MoveToNamespaceResult.Failed;
+            }
+
+            // The move service expects a single position, not a full selection
+            var moveSpan = new TextSpan(container.FullSpan.Start, 0);
```

Comment:
It demands a 0-length span but not just a position instead? Should that be fixed on that side?

Reply:
I'll file an issue on myself to remove that restriction and add tests to the movetype service. Will update this comment with issue number once I file

Reply:
#34643